src/Storages/StorageDistributed.cpp

```diff
@@ -1735,16 +1752,36 @@ void StorageDistributed::flushClusterNodesAllData(ContextPtr local_context)
             directory_queues.push_back(node.second.directory_queue);
     }
 
-    bool need_flush = getDistributedSettingsRef().flush_on_detach;
-    if (!need_flush)
-        LOG_INFO(log, "Skip flushing data (due to flush_on_detach=0)");
+    if (getDistributedSettingsRef().flush_on_detach)
+    {
+        LOG_INFO(log, "Flushing pending INSERT blocks");
 
-    /// TODO: Maybe it should be executed in parallel
-    for (auto & node : directory_queues)
+        Stopwatch watch;
+        ThreadGroupPtr thread_group = CurrentThread::getGroup();
+        ThreadPool pool(CurrentMetrics::StorageDistributedThreads, CurrentMetrics::StorageDistributedThreadsActive, CurrentMetrics::StorageDistributedThreadsScheduled, directory_queues.size());
+        for (const auto & node : directory_queues)
+        {
+            pool.scheduleOrThrowOnError([node_to_flush = node, &thread_group]()
+            {
+                SCOPE_EXIT_SAFE(
```

Comment:
same as above, maybe use `scheduleFromThreadPool` and wait for futures here

Reply:
Make sense, thanks, done.